### PR TITLE
Checking if object before calling method on it

### PR DESCRIPTION
**Pull request by Elektordi**
Prevents "Call to a member function toArray() on a non-object" fatal error when generating an xml with strings containing class-names from your application.

### CODE REVIEW

**Comment by ADmad** [comment]
Looks reasonable.

**Comment by dereuromark** [comment]
👍

**Comment by thinkingmedia** [comment]
what about `if(is_callable([$input, 'toArray'])`?

I'm not sure if `$input` was not an object if it would raise an error. I don't think it will.

**Comment by Elektordi** [comment]
@thinkingmedia I saw the form "is_object && method_exists" being used in others parts of cakephp, like in src/Utility/Hash.php for example.

**Comment by thinkingmedia** [comment]
@Elektordi oh if I could fix all the code in the code. I would!

You're change is okay, and it'll work in most cases but it's going to fail when `$input::toArray()` is **private** or **protected**.

That's because `method_exists` returns *true* no matter the accessibility of the method.

The `is_callable` method is safe as it will return true if you can not call the target in the current context (i.e. target could be a method, closure, etc.. etc..)

**Comment by thinkingmedia** [comment]
I would go with something like this.

```
        if (is_callable([$input, 'toArray']) {
              $input = call_user_func([$input, 'toArray']);
        }
```

Now, that's going to always work as expected.

It's these little things in PHP that make you a PHP ninja, or it makes you hate PHP. Depends on your perspective I guess.

**Comment by Elektordi** [comment]
Ok, you want me to update my branch with is_callable?

**Comment by thinkingmedia** [comment]
@Elektordi yes please.

**Comment by thinkingmedia** [comment]
Here's a good read on the topic.

http://jmfeurprier.com/2010/01/03/method_exists-vs-is_callable/

Looks like the recomend approach is to use both

```
   if(method_exists(...) && is_callable(..))
```

Seems that `is_callable` will always return `true` when the object implements `__call()` magic. So please use both.

**Comment by Elektordi** [comment]
I just tested it in my application, it doesn't work.
If input is a valid object name (but not the object itself), method_exists will return true, but then call_user_func will call the non-static function in a static way! (I didn't even know it was possible)
So in my case I got a "Using $this when not in object context" fatal error.

Maybe the best is to do something like: if(is_object(...) && method_exists(...) && is_callable(...))

**Comment by thinkingmedia** [comment]
Welcome to the world of PHP sadness.

**Comment by dereuromark** [comment]
Tests are failing now
https://travis-ci.org/cakephp/cakephp/jobs/175560130

Before it was a bugfix (just limiting it down a bit more), now it seems to be more a feature (allowing more like callables etc).

**Comment by Elektordi** [comment]
I'll install phpunit on my side and try to debug that...

**Comment by thinkingmedia** [comment]
@derekperkins and the sadness begins.. lol

I ran into something like this on Friday. I found this small piece of code in the core for Table.

```
    public function update($table)
    {
        $this->_dirty();
        $this->_type = 'update';
        $this->_parts['update'][0] = $table;

        return $this;
    }
```

So I changed `$this->_parts['update'][0] = $table;` to this `$this->_parts['update'] = (array)$table;`.

It's the same thing right? I mean, what harm could there be.

EVERYTHING EXPLODED.

I spent 3 hours trying to figure out all the special edge cases where that method was being called. The phpdocs says the argument is just a string type, but that's not have it's being used.

Thus I conclude. PHP sadness.

http://phpsadness.com/

**Comment by Elektordi** [comment]
Sorry, I was too quick with my edit, and it seems my test app did not use concerned code.
At least now I understand the importance of unit testing! :)

File: src/Utility/Xml.php [hunk 1]
```diff
@@ -203,8 +203,8 @@ protected static function _loadXml($input, $options)
      */
     public static function fromArray($input, $options = [])
     {
-        if (method_exists($input, 'toArray')) {
-            $input = $input->toArray();
+        if (is_object($input) && method_exists($input, 'toArray') && is_callable([$input, 'toArray'])) {
```
Comment (markstory) [inline src/Utility/Xml.php]: Is the `is_object()` check necessary? `method_exists()` behaves correctly for scalar values.

```php
php > var_dump(method_exists('nop', 'toArray'));
php shell code:1:
bool(false)
php > var_dump(method_exists(1.1, 'toArray'));
php shell code:1:
bool(false)
```

**Comment by Elektordi** [comment]
It only happens when the string is also a valid class name!
(Happens in my case because someone created an object in my db with the same name of a class in my model, and I have a xml export somewhere in a model object...)

Here is a quick test on how to trigger the bug:

```
<?php

include "cakephp/src/Utility/Xml.php";
include "cakephp/src/Utility/Exception/XmlException.php";
use Cake\Utility\Xml;

class TestClass {
    private $a = ['a'=>1,'b'=>2,'c'=>3];
    
    public function toArray() {
        return $this->a;
    }
}

echo "\tWith object:\n";
$c = new TestClass();
$x = Xml::fromArray(array('test'=>$c));
echo $x->asXML();

echo "\tWith string:\n";
$c = 'Toto';
$x = Xml::fromArray(array('test'=>$c));
echo $x->asXML();

echo "\tWith string which is also a class name:\n";
$c = 'TestClass';
$x = Xml::fromArray(array('test'=>$c));
echo $x->asXML();

echo "\n";
```

Output expected and with proposed PR:
```
	With object:
<?xml version="1.0" encoding="ISO-8859-1"?>
<test><a>1</a><b>2</b><c>3</c></test>
	With string:
<?xml version="1.0" encoding="ISO-8859-1"?>
<test>Toto</test>
	With string which is also a class name:
<?xml version="1.0" encoding="ISO-8859-1"?>
<test>TestClass</test>

```

Output with current master branch:
```
	With object:
<?xml version="1.0" encoding="ISO-8859-1"?>
<test><a>1</a><b>2</b><c>3</c></test>
	With string:
<?xml version="1.0" encoding="ISO-8859-1"?>
<test>Toto</test>
	With string which is also a class name:
PHP Fatal error:  Call to a member function toArray() on a non-object in /home/elektordi/src/cakephp-dev/cakephp/src/Utility/Xml.php on line 261
```

**Comment by markstory** [comment]
@Elektordi Ok. Is there a way we could get a test to help prevent future regressions around this?

**Comment by Elektordi** [comment]
Ok, got tired of this. I'm rewriting my branch to have a clean commit history...

**Pull request merged**